x/dex/types/shares.go

```diff
@@ -2,10 +2,76 @@ package types
 
 import (
 	"errors"
+	math "math"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
+/*
+For 2 asset pools, swap first to maximize the amount of tokens deposited in the pool.
+A user can deposit either one or 2 tokens, and we will swap first the biggest individual share and then join the pool.
+
+args:
+  - tokensIn: the tokens to add to the pool
+
+ret:
+  - numShares: the number of LP shares given to the user for the deposit
+  - remCoins: the number of coins remaining after the deposit
+  - err: error if any
+*/
+func (pool *Pool) SwapForSwapAndJoin(tokensIn sdk.Coins) (
+	out sdk.Coin, err error,
+) {
+	if len(pool.PoolAssets) != 2 {
+		err = errors.New("swap and add tokens to pool only available for 2 assets pool")
+		return
+	}
+
+	var x sdk.Int
+	var xDenom string
+	var y sdk.Int
+
+	// check who's x and y (x/)
+	if len(tokensIn) == 1 {
+		x = tokensIn[0].Amount
+		xDenom = tokensIn[0].Denom
+
+		y = sdk.ZeroInt()
+	} else {
+		// 2 assets
+		poolLiquidity := pool.PoolBalances()
+
+		s0 := tokensIn[0].Amount.ToDec().Quo(poolLiquidity.AmountOfNoDenomValidation(tokensIn[0].Denom).ToDec())
+		s1 := tokensIn[1].Amount.ToDec().Quo(poolLiquidity.AmountOfNoDenomValidation(tokensIn[1].Denom).ToDec())
+
+		if s0.GTE(s1) {
+			x = tokensIn[0].Amount
+			y = tokensIn[1].Amount
+
+			xDenom = tokensIn[0].Denom
+		} else {
+			x = tokensIn[1].Amount
+			y = tokensIn[0].Amount
+
+			xDenom = tokensIn[1].Denom
+		}
+	}
+
+	xIndex, xPoolAsset, err := pool.getPoolAssetAndIndex(xDenom)
+	lx := xPoolAsset.Token.Amount
+	ly := pool.PoolAssets[1-xIndex].Token.Amount
+
+	// x'=\sqrt{\frac{xk+kl_x}{y+l_y}}-l_x;\:x'=-\sqrt{\frac{xk+kl_x}{y+l_y}}-l_x
+	k := lx.Mul(ly)
+
+	xSwap := sdk.NewInt(
+		int64(math.Sqrt(
+			(x.Mul(k).Add(k.Mul(lx))).ToDec().Quo(
+				y.Add(ly).ToDec()).MustFloat64()))).Sub(lx)
+
+	return sdk.NewCoin(pool.PoolAssets[xIndex].Token.Denom, xSwap), err
```

Comment:
Can we use semantic names rather than letter representations here? It will be more verbose but easier to understand and keep clean.

Suggestions:
- `x` and `xDenom` → `xAmt`, `xDenom`
- `s0` and `s1` → `sharePctX`, `sharePctY` (percentages of the pool's x and y reserves owned)
- `k` → `swapInvariant` (as to not be confused for keeper)